### PR TITLE
Account memory of FileMetaData in global memory limit

### PR DESCRIPTION
**Pull request by hx235**
**Context/Summary:**
As revealed by heap profiling, allocation of `FileMetaData` for [newly created file added to a Version](https://github.com/facebook/rocksdb/pull/9924/files#diff-a6aa385940793f95a2c5b39cc670bd440c4547fa54fd44622f756382d5e47e43R774) can consume significant heap memory. This PR is to account that toward our global memory limit based on block cache capacity.

**Test plan:**
- Previous `make check` verified there are only 2 places where the memory of  the allocated `FileMetaData` can be released
- New unit test `TEST_P(ChargeFileMetadataTestWithParam, Basic)`
- db bench (CPU cost of `charge_file_metadata` in write and compact)
   - **write micros/op: -0.24%** : `TEST_TMPDIR=/dev/shm/testdb ./db_bench -benchmarks=fillseq -db=$TEST_TMPDIR -charge_file_metadata=1 (remove this option for pre-PR) -disable_auto_compactions=1 -write_buffer_size=100000 -num=4000000 | egrep 'fillseq'` 
   - **compact micros/op -0.87%** : `TEST_TMPDIR=/dev/shm/testdb ./db_bench -benchmarks=fillseq -db=$TEST_TMPDIR -charge_file_metadata=1 -disable_auto_compactions=1 -write_buffer_size=100000 -num=4000000 -numdistinct=1000 && ./db_bench -benchmarks=compact -db=$TEST_TMPDIR -use_existing_db=1 -charge_file_metadata=1 -disable_auto_compactions=1 | egrep 'compact'`
   
table 1 - write

#-run | (pre-PR) avg micros/op | std micros/op | (post-PR)  micros/op | std micros/op | change (%)
-- | -- | -- | -- | -- | --
10 | 3.9711 | 0.264408 | 3.9914 | 0.254563 | 0.5111933721
20 | 3.83905 | 0.0664488 | 3.8251 | 0.0695456 | -0.3633711465
40 | 3.86625 | 0.136669 | 3.8867 | 0.143765 | 0.5289363078
80 | 3.87828 | 0.119007 | 3.86791 | 0.115674 | **-0.2673865734**
160 | 3.87677 | 0.162231 | 3.86739 | 0.16663 | **-0.2419539978**

table 2 - compact

#-run | (pre-PR) avg micros/op | std micros/op | (post-PR)  micros/op | std micros/op | change (%)
-- | -- | -- | -- | -- | --
10 | 2,399,650.00 | 96,375.80 | 2,359,537.00 | 53,243.60 | -1.67
20 | 2,410,480.00 | 89,988.00 | 2,433,580.00 | 91,121.20 | 0.96
40 | 2.41E+06 | 121811 | 2.39E+06 | 131525 | **-0.96**
80 | 2.40E+06 | 134503 | 2.39E+06 | 108799 | **-0.78**

- stress test: `python3 tools/db_crashtest.py blackbox --charge_file_metadata=1  --cache_size=1` killed as normal

### CODE REVIEW

File: cache/cache_reservation_manager.h
Comment (hx235) [inline cache/cache_reservation_manager.h:39]: Question:
One annoying thing about the previous `UpdateCacheReservation()` is that it's not multi-thread friendly for `UpdateCacheReservation(GetTotalMemoryUsed() + delta)` as the "update" and the "get memory" is not atomic operation. 

Do you see a way of solving this other than open up another API?

**Comment by hx235** [comment]
Just CI - not ready for review.

**Comment by hx235** [comment]
Need to look into a more frequent crash_test failure  https://app.circleci.com/pipelines/github/facebook/rocksdb/13248/workflows/a5bb1b3c-f50d-4439-9ddd-d496c0f3ac70/jobs/358434 with this PR to see whether it's relevant.

**Comment by hx235** [comment]
> Need to look into a more frequent crash_test failure https://app.circleci.com/pipelines/github/facebook/rocksdb/13248/workflows/a5bb1b3c-f50d-4439-9ddd-d496c0f3ac70/jobs/358434 with this PR to see whether it's relevant.
> 
> But other than that, it's ready for review @ajkr !

Run the same command on upsteram/main with the same errors - so the db_stress failure is not related to this PR.

File: db_stress_tool/no_batched_ops_stress.cc
Comment (hx235) [inline db_stress_tool/no_batched_ops_stress.cc:13]: this is introduced by auto-format

File: java/src/main/java/org/rocksdb/RocksDB.java
Comment (hx235) [inline java/src/main/java/org/rocksdb/RocksDB.java:8]: Introduced by auto formatting

**Comment by hx235** [comment]
@ajkr ready for review again.

**Comment by ajkr** [comment]
> @ajkr ready for review again.

Sorry, since GitHub doesn't support stacked PRs I can only review one at a time without duplicate reviewing the same code, so am planning to wait until the dependency lands.

**Comment by hx235** [comment]
> > @ajkr ready for review again.
> 
> Sorry, since GitHub doesn't support stacked PRs I can only review one at a time without duplicate reviewing the same code, so am planning to wait until the dependency lands.

Not a problem - I need to do one more CI check too. I will request again when it's ready. Thanks!

File: db/dbformat.h
Comment (hx235) [inline db/dbformat.h:323]: Note: Needed for `size_t ApproximateMemoryUsage() const` in [db/version_edit.h](https://github.com/facebook/rocksdb/pull/9924/files#diff-1c2062fea938260158f1dfe0edd3889482515dc58e8b61e181df1a0329453a7d) below

**Comment by hx235** [comment]
@ajkr  after the CI signals are clear, it will be ready for review

**Comment by hx235** [comment]
Rebase

**Comment by hx235** [comment]
Fix some format

**Review by ajkr — APPROVED** [review]
LGTM!!

File: cache/cache_reservation_manager.h
Comment (hx235) [reply cache/cache_reservation_manager.h:39]: Let me look into this later.

Comment (hx235) [reply cache/cache_reservation_manager.h:39]: I like the "Alternatively you could introduce one more CacheReservationManager: a ConcurrentCacheReservationManager" better but this requires changes worth another small PR. Let me TODO and circle back to this during tech debt week.

Comment (hx235) [reply cache/cache_reservation_manager.h:43]: Done

File: HISTORY.md
Comment (hx235) [reply HISTORY.md]: Done  - also fixed the table.h

**Comment by hx235** [comment]
Update:
- Rebased and mainly addressed the conflict from https://github.com/facebook/rocksdb/pull/10040
- Addressed comments

File: cache/cache_reservation_manager.h
Comment (ajkr) [reply cache/cache_reservation_manager.h:43]: Sounds good. Yeah it's probably better to land this this week in time for the release than to block on that cleanup change.

**Pull request closed without merging**
